### ISSUE TITLE
Extend list of known Referrer Spammers

### ISSUE THREAD

**Issue by mattab** [piwik/piwik#5099]
In #2268 we have implemented a Referrer spam list, initially seeded with the worst of all spammers: semalt.

However it turns out there are thousands of other spammers that attack Piwik users with their lame websites. It will be hard to keep track of them all and write the list [in the config file here](https://github.com/piwik/piwik/blob/master/config/global.ini.php#L507-L510). 

List of spammers:
* see the current list of referrer spammer [in Piwik global.ini.php file here](https://github.com/search?utf8=%E2%9C%93&q=referrer_urls_spam+user%3Apiwik+path%3Aconfig&type=Code&ref=searchresults)

What is our best way to move forward?

**Comment by ranzeflitze** [comment]
@mattab This is what I pasted into the config file (at the end of the file after all the PluginsInstalled)
> [Tracker]
; Comma separated list of known Referrer Spammers, ie. bot visits that set a fake Referrer field.
; All Visits with a Referrer URL host set to one of these will be excluded.
; If you find new spam entries in Referrers>Websites, please report them here: https://github.com/piwik/piwik/issues/5099
referrer_urls_spam = "semalt.com,buttons-for-website.com,make-money-online.7makemoneyonline.com,7makemoneyonline.com”

Thank you.

**Comment by Trance-Man** [comment]
streamcyclone.com
perkinslawgroup.net

**Comment by fvdm** [comment]
+ anticrawler.org
Interesting, they are to prevent bots or so they say... probably going to serve malware.

**Comment by PeterTheOne** [comment]
+1 anticrawler.org

**Comment by ways2web** [comment]
abcd4.de
findsimilarsites.de

**Comment by samwaterston** [comment]
+1 for anticrawler.org. Looks like it's Semalt again: http://www.cradlecloud.com/how-to-ban-and-block-anticrawler-org-spam-referrals/

**Comment by cdtoad** [comment]
Just got hit with about 50 buttons-for-website.com referrer in my dashboard.  Upgraded yesterday to 2.10.0 and do see

**Comment by fvdm** [comment]
@cdtoad there is no toggle, would be nice though..

**Comment by brynnd** [comment]
buttons-for-website seems to be associated with http://sharebutton.net/

**Comment by brynnd** [comment]
netvibes.com

**Comment by brynnd** [comment]
ranksonic.info  Not sure if this is spam or not.  But looks like to me.

**Comment by samwaterston** [comment]
+1 for ranksonic.info

**Comment by brynnd** [comment]
I wonder if there has been any decision made about how to handle this issue.  I've been putting off adding new spammers to the block list myself, and instead, wait for the next upgrade.  But buttons-for-website and ranksonic have become prolific, and waiting for the next upgrade will not be possible for me.  Now I'm going to follow the instructions that are posted above, and add them myself.  But a nice button click would be very nice!  Thanks  :-)

**Comment by brynnd** [comment]
Well, it turns out I'm going to need a little help, to add them myself.  Up near the top of this thread, a comment by jlj says to edit this file:  piwik/config/global.ini.php.  And then it shows where to add the spam urls.

(Comma separated list of known Referrer Spammers, ie. bot visits that set a fake Referrer field.
All Visits with a Referrer URL host set to one of these will be excluded.
If you find new spam entries in Referrers>Websites, please report them here: https://github.com/piwik/piwik/issues/5099
referrer_urls_spam = "semalt.com,savetubevideo.com,kambasoft.com")

But when I find the global.ini.php file, it says near the top "; WARNING - YOU SHOULD NOT EDIT THIS FILE DIRECTLY - Edit config.ini.php instead."  However, when I open config.ini.php, it doesn't look anything like global.ini.php.  It doesn't even have the code from global.ini.php, where ljl says I need to add the spam urls.

Later in this thread (about halfway down the page) scheinercc gives some instructions with a list of all the urls to add.  His instructions say:
"
    in piwik config file piwik/config/config.ini.php
    add below [Tracker] section:
"
and then he gives the same code that ljl indicated from the global.ini.php file.  However, [Tracker] does not appear in config.ini.php.

So I am pretty thoroughly confused.  Could someone clear this up for me?  Where do I add the referrer spam urls?

Thank you very much  :-)

**Comment by fvdm** [comment]
@brynnd don't directly change `global.ini.php`, it's only for the default settings and may be reset when upgrading Piwik. Instead copy the parts you wish to change to `config.ini.php`. When the relevant section like `[Tracker]` is missing you simply add it yourself:

```ini
[Tracker]
referrer_urls_spam = "semalt.com,savetubevideo.com,kambasoft.com"
```

**Comment by jlj** [comment]
@brynnd I just edited the comment above to clarify which config file shall be edited for future readers of this thread. ;-)

**Comment by brynnd** [comment]
So I don't need to add the part that starts with ";Comma separated list...."?  I just need to put

[Tracker]
referrer_urls_spam = "semalt.com,savetubevideo.com,kambasoft.com"

in config.ini.php?  Do I just put it at the end of config.ini.php?

**Comment by fvdm** [comment]
@brynnd The lines that start with a `;` are comments and therefore ignored by Piwik. You can add them for your own (future) reference, but they are not required.

If the `[Tracker]` section does not exist yet it's fine to add it to the end of the `config.ini.php` file.

**Comment by brynnd** [comment]
Ok, thanks fvdm  :-)

I've edited the file, and by tomorrow, I should know if it's working.  The buttons-for-website and ranksonic have been coming every day lately.  Although buttons-for-website first came up long ago enough that it should be in my version of piwik.  So since these "manual install" instructions were posted, maybe they haven't been keeping the list up to date, in new versions?

Here's an updated list (comma-separated), since the list provided by scheinercc hasn't been updated lately.  Maybe we could keep it updated, with the addition of each new domain/URL?

Maybe every time someone reports a new domain/URL, they can copy this list, paste it in their message, and add their new item to it.  If everyone does that, we can have an always current and complete list?

Just a thought  :-)

5.kambasoft.com/2.php,9.kambasoft.com/2.php,16782868.website-errors-scanner.com,39946554.semalt.com/crawler.php,7makemoneyonline.com,abcd4.de,anticrawler.org,bodywisesupplements.com,buttons-for-website.com,caldercarrentals.com,coastalstickers.com.au,findsimilarsites.de,herahair.com,icompbusiness.com,livefixer.com,make-money-online.7makemoneyonline.com,make-money-online.7makemoneyonline.com/money.php,maranbrinfo.com.br,mountainstream.ms,musica.descargar-musica-gratis.net/descargar-musica-gratis.php,musicas.baixar-musicas-gratis.com/baixar-musicas-gratis.php,musicas.kambasoft.com/2.php,netvibes.com,perkinslawgroup.net,ranksonic.info,religionenserio.com,semalt.com,semalt.com/crawler.php,semalt.semalt.com/crawler.php,seo.my-api.com,sharebutton.net,speechtotextservice.com,streamcyclone.com,textelle.ee,ticimax.com,waterfallscanopy.com,www.semalt.com,www.semalt.com/crawler.php,www.star61.de,www.windstream.net,youtube-downloader.savetubevideo.com/youtube-downloader.php

**Comment by fvdm** [comment]
You don't have to write full urls. Piwik checks if the referrer _contains_ the string, so only including `7makemoneyonline.com` is enough for all urls that mention this part somewhere.

**Comment by mattab** [comment]
Hi guys, i'm adding the following to the list of referrer spammer: `anticrawler.org,ranksonic.info,savetubevideo.com,kambasoft.com` - if there are more, please let me know

**Comment by larsactionhero** [comment]
Hi there,

I've created a alphabetically ordered list (just to make it more comfortable), maybe I forgot some entries, maybe some of them are not neccessary, I'm not sure.
Feel free to edit and repost :)


<code>7makemoneyonline.com,7makemoneyonline.com,7makemoneyonline.com,abcd4.de,anticrawler.org,anticrawler.org,bodywisesupplements.com,buttons-for-website.com,caldercarrentals.com,coastalstickers.com.au,descargar-musica-gratis.net,findsimilarsites.de,findsimilarsites.de,herahair.com,herahair.com,icompbusiness.com,kambasoft.com,livefixer.com,maranbrinfo.com.br,mountainstream.ms,musicas.baixar-musicas-gratis.com,netvibes.com,perkinslawgroup.net,perkinslawgroup.net,ranksonic.info,ranksonic.info,religionenserio.com,savetubevideo.com,seo.my-api.com,sharebutton.net,semalt.com,speechtotextservice.com,speechtotextservice.com,star61.de,streamcyclone.com,streamcyclone.com,textelle.ee,ticimax.com,urlopener.blogspot.com,waterfallscanopy.com,website-errors-scanner.com,windstream.net,youtube-downloader.savetubevideo.com,7makemoneyonline.com,livefixer.com</code>

**Comment by cohan** [comment]
Switching from Google Analytics to piwik for this exact reason. Couple new ones I've started seeing that don't seem to be listed here already

o-o-6-o-o.com,bestwebsitesawards.com,*.darodar.com,ranksonic.org,ranksonic.info,delta-search.com,sr.searchfunmoods.com

**Comment by mattab** [comment]
Hi guys,

Thanks for the feedback. I'm adding the following websites that i've confirmed to be referrer spammers:
 * ilovevitaly.com, priceg.com, blackhatworth.com, hulfingtonpost.com
 * darodar.com, econom.co, o-o-6-o-o.com,bestwebsitesawards.com, darodar.com,ranksonic.org,ranksonic.info

If you have others confirmed, please report them here!

**Comment by mattab** [comment]
> Switching from Google Analytics to piwik for this exact reason

That's the way to go @cohan :-)

**Comment by gfive** [comment]
What about http://www.projecthoneypot.org ? use it for my forum, email forms and others. Works like a charm...

**Comment by mattab** [comment]
@gfive does it provide a list of referrer spammers as well?

**Comment by gfive** [comment]
I think so. But main function is to get all spammers by IP. I used to get lot's of visits from semalt.com as well. But after installing my honeypot It completely stopped.
Account is free at http://www.projecthoneypot.org so you could register and check if it offers where you are looking for...

**Comment by brynnd** [comment]
I have a honey pot on my site, but it doesn't seem to stop the referrer spam.  I guess these referrer spammers aren't traditional spammers.

If you installed the honey pot around the time you upgraded Piwik, it might have been a coincidence.

**Comment by samwaterston** [comment]
Hi,

New spam referrer (I guess): 4webmasters.org referral traffic from Russia started showing 4 days ago on my website stats.

**Comment by brynnd** [comment]
best-seo-solutions.com
It's a semalt-related domain.

**Comment by garfieldairlines** [comment]
Same here : best-seo-solution.com all from Brazil from Thursday 2 April 2015 - 21:00 to Friday 3 April 2015 - 02:00 : 
189.58.107.27 
186.225.151.50 
201.42.91.49 
189.106.183.216

**Comment by Joey3000** [comment]
+1 for best-seo-solution.com

Additional note: referrer_urls_spam currently has following entries twice:
darodar.com
ranksonic.info

The entries could be sorted alphabetically, to prevent duplication in future.

**Comment by mattab** [comment]
best-seo-solution.com was added to the list, and the list was sorted in alphabetical order. We'll keep listening to your feedback in this issue :+1:

**Comment by AgentGod** [comment]
Hello to All
@mattab maybe the best way to keep the spammer list will be to keep them in external file instead in global.ini.
Option in administration panel Enable/Disable/ spammer list, plus button next to them to update the list, so it will be possible to update only the list. 
Another option is again Enable/Disable spammer list, plus text box next to them to easily add spam url.

**Comment by AgentGod** [comment]
Also one spammer from me humanorightswatch.org
In google analytics I have site13.simple-share-buttons.com and site11.simple-share-buttons.com
Will the entry for simple-share-buttons.com block also site13.simple-share-buttons.com, if not is it possible to add something like *.simple-share-buttons.com

**Comment by jazzcrack** [comment]
+1 best-seo-solutions.com

**Comment by Globulopolis** [comment]
Totally agree with @AgentGod. And I think support for regex in URL to reduce the list of similar URLs

**Comment by brynnd** [comment]
Note that there's a best-seo-solutions.com, and best-seo-solution.com!  Gggrrr!!

**Comment by gallolu** [comment]
Please add also buttons-for-your-website.com and best-seo-solution.com .. :)

**Comment by brynnd** [comment]
I guess this is a new trend, and perhaps needs to be considered in a permanent feature.  Now it looks like these guys are changing their domain names just slightly, to bypass our blocks!

I leads me to wonder (AGIAN!) what benefit these guys gain in doing this.  "Spam" is "spam" and I guess no one really knows or understands its purpose.  But I still do wonder why they waste time on this.  What benefit could there possibly be?  For anyone?

Note that also there is buttons-for-website.com and buttons-for-your-website.com  ggrrr!

Also, another new one:  best-seo-offer.com

**Comment by AgentGod** [comment]
+1 best-seo-offer.com
+1 buttons-for-your-website.com

**Comment by fvdm** [comment]
+1 for regex support, would be very useful

**Comment by campino2k** [comment]
+1 best-seo-offer.com

**Comment by mnapoli** [comment]
> I guess this is a new trend, and perhaps needs to be considered in a permanent feature. Now it looks like these guys are changing their domain names just slightly, to bypass our blocks!

I agree this is getting more and more of a problem. By the way it seems that Google Analytics is not tackling this problem yet (I see all those spammers in GA), which gives Piwik a nice little plus.

Anyway I believe too that we need to find a more robust solution for this: as a user I don't like to have to wait for new Piwik releases to exclude new spammers (my data keeps being polluted in the meantime). I also think the current way of reporting spammers (GitHub issue) is not the best.

We need:

- to make it easier for users to report new spammers
- Piwik to auto-update the spammers list
- while still keep the list up to date in new releases (for the Piwik installs that are setup to avoid any external network call)

@mattab I'd like to open a separate issue to discuss a solution (I have a few ideas) so that we keep this issue for reporting spammers, what do you think? Should we tackle this soon or let it be for now?

**Comment by mattab** [comment]
> @mattab I'd like to open a separate issue to discuss a solution (I have a few ideas) so that we keep this issue for reporting spammers, what do you think? Should we tackle this soon or let it be for now?

something we can try with low effort is to put the list in a separate file with one spammer per line, and then ask the community to issue pull request, because it's so easy to make a PR on github web interface alone, it could be quick and efficient solution. 

to discuss auto update feature +1 to discuss in a separate new issue to make sure it is followed up

**Comment by mnapoli** [comment]
I've opened #7674 to continue the discussion.

**Comment by mnapoli** [comment]
It seems we are not the only one trying to build such list, see http://www.reddit.com/r/Wordpress/comments/2qteln/i_want_to_build_a_list_of_referrer_spam_links_to/ (seems to be updated regularly)

**Comment by Fensterbank** [comment]
While there is no better way to report spammers at the moment, I'll continue here :)
* buttons-for-your-website.com (which is not the same than buttons-for-website.com, which is already added to the list)
* best-seo-offer.com

**Comment by AgentGod** [comment]
Entire list alphabetically: 

4webmasters.org,7makemoneyonline.com,adcash.com,anticrawler.org,best-seo-offer.com,best-seo-solution.com,bestwebsitesawards.com,blackhatworth.com,buttons-for-website.com,buttons-for-your-website.com,cenokos.ru,cenoval.ru,cityadspix.com,darodar.com,econom.co,iskalko.ru,edakgfvwql.ru,forum.smailik.org,Get-Free-Traffic-Now.com,gobongo.info,googlsucks.com,hulfingtonpost.com,humanorightswatch.org,ilovevitaly.co,ilovevitaly.com,ilovevitaly.ru,kambasoft.com,luxup.ru,make-money-online.7makemoneyonline.com,myftpupload.com,o-o-6-o-o.ru,o-o-8-o-o.ru,priceg.com,prlog.ru,ranksonic.info,ranksonic.org,savetubevideo.com,screentoolkit.com,semalt.com,semalt.semalt.com,seoexperimenty.ru,simple-share-buttons.com,slftsdybbg.ru,social-buttons.com,socialseet.ru,superiends.org,theguardlan.com,vodkoved.ru,websocial.me,ykecwqlixx.ru

**Comment by adegans** [comment]
I think this can easily be countered if someone builds a thingy in Piwik to download a new list every so much hours, say once every 24 hours.

And the serving system where people can submit links. If 10 (?) people submit the same link it's added in the downloadable list which each piwik setup can fetch every 24 hours. This could employ a simple filter to get double submissions like www.semalt.com vs. semalt.com so the list stays somewhat compact and clean.

This submission system can be a thing in Github (Similar to torrent blocklists - https://gist.github.com/johntyree/3331662) or something more fancy. 

I'm not overly familiar with Github and new to Piwik. Just came across this because I have a referrer spam problem also.

**Comment by mnapoli** [comment]
The list has been moved to https://github.com/piwik/referrer-spam-blacklist in order to make it more visible and more practical to maintain.

*Please open issues and pull requests in that new repository :)*

If you are interested to know how this list will be handled, read this issue: #7674

@Fensterbank thanks I confirm those 2, I have added them to the new list.

**Comment by mattab** [comment]
## How do I contribute to the Referrer spammer Piwik list?

To add a new referrer spammer to the list, [click here to edit the spammers.txt file](https://github.com/piwik/referrer-spam-blacklist/edit/master/spammers.txt) and create a pull request. Alternatively you can create a [new issue](https://github.com/piwik/referrer-spam-blacklist/issues/new).  

Looking forward in the  future to maintaining this spammer list together as a community :+1:

**Comment by sbrickey** [comment]
Since the list is now maintained via file on github, any chance that the admin site can check for newer versions, and ideally go download it automatically?

**Comment by mnapoli** [comment]
@sbrickey see #7674

**Comment by CgColorsArizona** [comment]
Check this in-depth blog to resolve this issue...
<a href="http://www.cgcolors.com/blog/block-unwanted-referral-traffic-google-analytics-report/">How to block Google analytics' spamming (nonhuman) referral sites</a>

I think need to action by Google Analytics' team.

**Comment by brynnd** [comment]
Well, it's exciting to see some new energy and seeing this project move forwards!

Unfortunately, I'm not very familiar with programming or how this site works.  When I click the link in mattab's last msg, to edit the spammers.txt file, it says:

-- You need to fork this repository to propose changes.
-- Sorry, you’re not able to edit this repository directly— you need to fork it and propose your changes from there instead. 

I don't know what that means to "fork this repository".  So I used the other option and reported a new spammer (another semalt variety).  But I'm still not sure how to proceed with my piwik installation.

I did read #7674, but unfortunately, I don't understand much of it.  I also read https://github.com/piwik/referrer-spam-blacklist, but again, don't understand much.

If I have the current version of Piwik, am I getting all the spammers blocked?  Or do I need to continue adding new spammers to my config.ini.php file?  

Thank you very much  :-)

**Comment by mnapoli** [comment]
@brynnd it's fine if you weren't able to edit the file directly, opening an issue is good too. When it is added to the list it will be included in the **new** Piwik version, so that's why it's important to keep Piwik up to date.

In the future we want to auto-update the list so that you get the latest spammers blocked even before the new Piwik release is available (issue #7674).

**Comment by brynnd** [comment]
Thanks mnapoli!

Where can I check the current list, so I don't accidentally add duplicates to the list?  Especially these new semalt-related one, where they're just changing the domain by a character or 2/

**Comment by mattab** [comment]
@brynnd latest version is at: https://github.com/piwik/referrer-spam-blacklist/

Please note: you don't need to add the semalt variation if they are sub-domains of semalt.com (or any other spammer). but if they are new domain names (not sub-domains) then please suggest the new spammers on this project: https://github.com/piwik/referrer-spam-blacklist/

**Comment by tombrossman** [comment]
I've added several domains to piwik/config/config.ini.php as described above. A week has passed and every day I see new entries from these same domains. Am I missing something basic, or is this worth opening a new issue?

This is at the end of my config.ini.php, and I've restarted Apache (and later the server) but modifying this configuration file has had no effect at all:
```
[Tracker]
referrer_urls_spam = "100dollars-seo.com,semaltmedia.com"
```

**Comment by mnapoli** [comment]
@tombrossman with the latest Piwik versions this INI config option isn't used anymore. There will be a new Piwik release very soon (probably tomorrow), else you can update to the latest beta and those spammers will be blocked.

With Piwik 2.14 the spammers list will be updated automatically.

**Comment by tombrossman** [comment]
Ah, thanks - that makes sense now. I thought it was me doing something stupid again...

**Comment by jpjp** [comment]
Is anti-referral spamming included in piwik 2.14 and enabled by default? Is there any way to retroactively apply it, like with the geoip location dbs? I couldn't find any documentation on this. Thanks.

**Comment by mnapoli** [comment]
> Is anti-referral spamming included in piwik 2.14 and enabled by default?

Yes

> Is there any way to retroactively apply it, like with the geoip location dbs?

No, it doesn't apply retroactively.
